manager.cpp

```diff
@@ -117,37 +118,104 @@ void Manager::driveLEDs(group& ledsAssert, group& ledsDeAssert)
-    if (ledsDeAssert.size())
+    group newReqChangedLeds;
+    std::vector<std::pair<group&, group&>> actionsVec = {
+        {reqLedsAssert, ledsAssert}, {reqLedsDeAssert, ledsDeAssert}};
+
+    timer.setEnabled(false);
+    std::set_union(ledsAssert.begin(), ledsAssert.end(), ledsDeAssert.begin(),
+                   ledsDeAssert.end(),
+                   std::inserter(newReqChangedLeds, newReqChangedLeds.begin()),
+                   ledLess);
+
+    // prepare reqLedsAssert & reqLedsDeAssert
+    for (auto pair : actionsVec)
     {
-        for (const auto& it : ledsDeAssert)
-        {
-            std::string objPath = std::string(PHY_LED_PATH) + it.name;
-            lg2::debug("De-Asserting LED, NAME = {NAME}", "NAME", it.name);
-            drivePhysicalLED(objPath, Layout::Action::Off, it.dutyOn,
-                             it.period);
-        }
+        group tmpSet;
+
+        // Discard current required LED actions, if these LEDs have new actions
+        // in newReqChangedLeds.
+        std::set_difference(pair.first.begin(), pair.first.end(),
+                            newReqChangedLeds.begin(), newReqChangedLeds.end(),
+                            std::inserter(tmpSet, tmpSet.begin()), ledLess);
+
+        // Union the remaining LED actions with new LED actions.
+        pair.first.clear();
+        std::set_union(tmpSet.begin(), tmpSet.end(), pair.second.begin(),
+                       pair.second.end(),
+                       std::inserter(pair.first, pair.first.begin()), ledLess);
     }
 
-    if (ledsAssert.size())
+    driveLedsHandler();
+    return;
+}
+
+static bool isPSUObjPathValid(const std::string& objPath)
+{
+    if (invalidLedGroups.find(objPath) == invalidLedGroups.end())
     {
-        for (const auto& it : ledsAssert)
-        {
-            std::string objPath = std::string(PHY_LED_PATH) + it.name;
-            lg2::debug("Asserting LED, NAME = {NAME}", "NAME", it.name);
-            drivePhysicalLED(objPath, it.action, it.dutyOn, it.period);
-        }
+        return true;
     }
-    return;
+    return false;
+}
+
+static bool isSysfsPresentForPSU(const std::string& objPath1)
+{
+    auto search1 = "cffps1";
+    auto search2 = "cffps2";
+
+    auto ledGroup1 = objPath1.substr(objPath1.rfind('/') + 1);
+    std::replace(ledGroup1.begin(), ledGroup1.end(), '_', '-');
+
+    auto ledGroup2 = ledGroup1;
+    auto objPath2 = objPath1;
+
+    if (ledGroup2.find(search1) != std::string::npos)
+    {
+        ledGroup2.replace(ledGroup2.begin(), ledGroup2.begin() + 6, search2);
+        objPath2.replace(objPath2.rfind('/') + 1, 6, search2);
+    }
+    else if (ledGroup2.find(search2) != std::string::npos)
+    {
+        ledGroup2.replace(ledGroup2.begin(), ledGroup2.begin() + 6, search1);
+        objPath2.replace(objPath2.rfind('/') + 1, 6, search1);
+    }
+
+    auto sysLedsPrefix = "/sys/class/leds/";
+    std::string path1 = sysLedsPrefix + ledGroup1;
```

Comment:
Do we actually need path1 and path2 or values can be used in place?
Also get rid of path1 and path2... some meaningful names :)